File: telemetry/aitf/sdk/python/src/aitf/ocsf/schema.py
"""AITF OCSF Base Schema.

OCSF v1.9.0 base objects and AI-specific extension models.
Based on the OCSF schema from the AITelemetry project, enhanced
for AITF AI events under the OCSF class-reuse model.

OCSF v1.9.0 (2026-08-03) landed the ``ai_agent`` and ``delegation`` objects and
extended the ``ai_operation`` profile onto the ``system``, ``network``,
``application`` and ``iam`` base classes, so every class in those categories
inherits agent attribution. AITF therefore emits *only* existing OCSF classes;
there is no bespoke AI category. See ``spec/ocsf-mapping/event-classes.md``.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import IntEnum
from typing import Any

from pydantic import BaseModel, Field, model_validator


# --- OCSF Enumerations ---

class OCSFSeverity(IntEnum):
    UNKNOWN = 0
    INFORMATIONAL = 1
    LOW = 2
    MEDIUM = 3
    HIGH = 4
    CRITICAL = 5
    FATAL = 6


class OCSFStatus(IntEnum):
    UNKNOWN = 0
    SUCCESS = 1
    FAILURE = 2
    OTHER = 99


class OCSFActivity(IntEnum):
    UNKNOWN = 0
    CREATE = 1
    READ = 2
    UPDATE = 3
    DELETE = 4
    OTHER = 99


class OCSFCategoryUID(IntEnum):
    """OCSF category UIDs that AITF AI events map onto.

    Every member is a **released** OCSF category. Following OCSF's "reuse
    existing objects and profiles" direction (PR #1641 / issue #1640), AITF
    emits AI telemetry under existing categories enriched with the
    ``ai_operation`` profile, which OCSF v1.9.0 attached to the ``system``,
    ``network``, ``application`` and ``iam`` base classes.

    There is deliberately no ``AI`` member. OCSF ``categories.json`` stops at
    uid 8 (``unmanned_systems``); a dedicated AI category has not been
    ratified, so AITF does not emit one. See ``LEGACY_AI_CLASS_UIDS``.
    """
    FINDINGS = 2
    IAM = 3
    DISCOVERY = 5
    APPLICATION = 6


class OCSFClassUID(IntEnum):
    """OCSF event class UIDs that AITF AI events map onto.

    All members are released OCSF classes. Agent, delegation and agent-to-agent
    communication lifecycle — previously modelled as provisional classes 9001,
    9002 and 9003 — now reuse ``API_ACTIVITY`` and ``AUTHORIZE_SESSION``,
    carrying AI context on the ``ai_operation`` profile.
    """
    # Reused existing OCSF classes (verified against OCSF v1.9.0).
    VULNERABILITY_FINDING = 2002
    COMPLIANCE_FINDING = 2003
    DETECTION_FINDING = 2004
    ACCOUNT_CHANGE = 3001
    AUTHENTICATION = 3002
    AUTHORIZE_SESSION = 3003
    ENTITY_MANAGEMENT = 3004
    USER_ACCESS_MANAGEMENT = 3005
    INVENTORY_INFO = 5001
    WEB_RESOURCES_ACTIVITY = 6001
    APPLICATION_LIFECYCLE = 6002
    API_ACTIVITY = 6003
    DATASTORE_ACTIVITY = 6005


# Backward-compatible alias. AITF previously defined a bespoke Category 7 with
# classes 7001-7010; events now reuse the OCSF classes above per OCSF's
# object/profile-reuse model. Kept so existing imports keep working.
AIClassUID = OCSFClassUID


# Retired AITF class UIDs -> the released OCSF class each now maps onto.
#
# AITF briefly emitted three provisional classes in a proposed ``ai`` category
# (uid 9). That category was never ratified — OCSF v1.9.0 ``categories.json``
# still ends at uid 8 — and v1.9.0 instead solved agent attribution by putting
# ``ai_agent`` and ``delegation`` on the ``ai_operation`` profile and attaching
# that profile to the application and IAM base classes. The provisional classes
# are therefore retired. This table exists only so consumers can decode
# telemetry recorded by AITF <= 0.4 during the proposal period.
LEGACY_AI_CLASS_UIDS: dict[int, int] = {
    9001: OCSFClassUID.API_ACTIVITY,        # agent_activity
    9002: OCSFClassUID.AUTHORIZE_SESSION,   # delegation_activity
    9003: OCSFClassUID.API_ACTIVITY,        # agent_communication
}


class AgentTypeID(IntEnum):
    """OCSF ``ai_agent.type_id`` — normalized agent framework.

    Matches ``objects/ai_agent.json`` as **released in OCSF v1.9.0**: the six
    members below are the full upstream enum. Note that the ``MCP`` and ``A2A``
    members floated during the PR #1641 discussion did not ship; protocol is
    modelled on the message, not on the agent's framework type.
    """
    UNKNOWN = 0
    NATIVE = 1
    LANGCHAIN = 2
    AUTOGEN = 3
    CREWAI = 4
    OTHER = 99


# Caption labels for AgentTypeID, per OCSF PR #1641.
AGENT_TYPE_LABELS: dict[int, str] = {
    0: "Unknown",
    1: "Native",
    2: "LangChain",
    3: "AutoGen",
    4: "CrewAI",
    99: "Other",
}

# AITF framework value -> OCSF ai_agent.type_id. Frameworks without a
# dedicated OCSF enum member (langgraph, semantic_kernel, custom, ...)
# normalize to OTHER (99), matching OCSF's open-enum guidance.
_FRAMEWORK_TO_TYPE_ID: dict[str, int] = {
    "native": AgentTypeID.NATIVE,
    "langchain": AgentTypeID.LANGCHAIN,
    "langgraph": AgentTypeID.LANGCHAIN,
    "autogen": AgentTypeID.AUTOGEN,
    "crewai": AgentTypeID.CREWAI,
}


def normalize_agent_type_id(framework: str | None) -> int:
    """Map an AITF framework string to an OCSF ``ai_agent.type_id`` value."""
    if not framework:
        return AgentTypeID.UNKNOWN
    return int(_FRAMEWORK_TO_TYPE_ID.get(framework.strip().lower(), AgentTypeID.OTHER))


# Retired. AITF no longer emits a bespoke AI category: OCSF v1.9.0 ships no
# such category (``categories.json`` ends at uid 8) and instead carries agent
# and delegation context on the ``ai_operation`` profile. Retained as ``None``
# so code that imported this constant fails loudly rather than silently
# emitting an unratified category_uid.
OCSF_AI_CATEGORY_UID: int | None = None


# --- OCSF Base Objects ---

class OCSFMetadata(BaseModel):
    """OCSF event metadata."""
    version: str = "1.9.0"
    product: dict[str, str] = Field(
        default_factory=lambda: {
            "name": "AITF",
            "vendor_name": "AITF",
            "version": "0.4.0",
        }
    )
    uid: str = Field(default_factory=lambda: str(uuid.uuid4()))
    correlation_uid: str | None = None
    original_time: str | None = None
    logged_time: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )


class OCSFActor(BaseModel):
    """OCSF actor information."""
    user: dict[str, Any] | None = None
    session: dict[str, Any] | None = None
    app_name: str | None = None


class OCSFDevice(BaseModel):
    """OCSF device/host information."""
    hostname: str | None = None
    ip: str | None = None
    type: str | None = None
    os: dict[str, str] | None = None
    cloud: dict[str, str] | None = None
    container: dict[str, str] | None = None


class OCSFEnrichment(BaseModel):
    """OCSF enrichment data."""
    name: str
    value: str
    type: str | None = None
    provider: str | None = None


class OCSFObservable(BaseModel):
    """OCSF observable value."""
    name: str
    type: str
    value: str


# --- AI-Specific Extension Models ---

class AIModelInfo(BaseModel):
    """AI model information."""
    model_id: str
    name: str | None = None
    version: str | None = None
    provider: str | None = None
    type: str | None = None  # llm, embedding, image, audio, multimodal
    parameters: dict[str, Any] | None = None


class AITokenUsage(BaseModel):
    """AI token usage statistics."""
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    cached_tokens: int = 0
    reasoning_tokens: int = 0
    estimated_cost_usd: float | None = None

    @model_validator(mode="after")
    def compute_total(self) -> "AITokenUsage":
        if self.total_tokens == 0:
            self.total_tokens = self.input_tokens + self.output_tokens
        return self


class AILatencyMetrics(BaseModel):
    """AI operation latency metrics."""
    total_ms: float = 0.0
    time_to_first_token_ms: float | None = None
    tokens_per_second: float | None = None
    queue_time_ms: float | None = None
    inference_time_ms: float | None = None


class AICostInfo(BaseModel):
    """AI operation cost information."""
    input_cost_usd: float = 0.0
    output_cost_usd: float = 0.0
    total_cost_usd: float = 0.0
    currency: str = "USD"


class AITeamInfo(BaseModel):
    """Multi-agent team information."""
    team_name: str
    team_id: str | None = None
    topology: str | None = None
    members: list[str] = Field(default_factory=list)
    coordinator: str | None = None


class AISecurityFinding(BaseModel):
    """Security finding details."""
    finding_type: str
    owasp_category: str | None = None
    risk_level: str
    risk_score: float
    confidence: float
    detection_method: str = "pattern"
    blocked: bool = False
    details: str | None = None
    pii_types: list[str] = Field(default_factory=list)
    matched_patterns: list[str] = Field(default_factory=list)
    remediation: str | None = None


class OCSFAIAgent(BaseModel):
    """OCSF ``ai_agent`` object — **released in OCSF v1.9.0**.

    An autonomous AI agent operating under delegated authority. Distinct from
    the OCSF ``agent`` object (which models security sensors such as EDR/DLP)
    and from human principals. Attached to events via the ``ai_operation``
    profile so any activity can be attributed to the agent that performed it.

    All eight fields below are upstream ``objects/ai_agent.json`` attributes;
    AITF adds nothing here. Since v1.9.0 attached ``ai_operation`` to the
    ``system``, ``network``, ``application`` and ``iam`` base classes, this
    object can ride on any event in those categories.
    """
    uid: str  # required: stable logical identifier
    instance_uid: str | None = None  # restart-sensitive running instance id
    name: str | None = None
    type: str | None = None  # caption of type_id (Native, LangChain, ...)
    type_id: int = AgentTypeID.UNKNOWN
    ai_model: str | None = None  # model backing the agent at event time
    version: str | None = None  # agent code/configuration revision
    charter: str | None = None  # role / operating-boundary reference


class OCSFDelegation(BaseModel):
    """OCSF ``delegation`` object — **released in OCSF v1.9.0**.

    A durable authorization context issued by a principal to a delegate,
    persisting independently of any single trace, session or workflow instance.
    Reached via the ``ai_operation`` profile, so all events performed under the
    same delegated authority correlate on ``uid``.

    The first four fields are the released upstream attributes. The rest are an
    **AITF extension** and are the remaining OCSF ask: upstream models *that* a
    delegation exists and who issued it, but not what authority it conveys.
    Scope attenuation and proof-of-possession are what make a delegation chain
    auditable, so AITF keeps emitting them pending upstream acceptance.
    """
    # --- released OCSF v1.9.0 attributes ---
    uid: str  # required: stable delegation identifier
    created_time: str | None = None  # when the issuing authority minted it
    issuer_uid: str | None = None  # trusted issuer that minted the delegation
    parent_uid: str | None = None  # parent delegation (re-delegation lineage)
    # --- AITF extension: pending upstream, see upstream-pr-plan-ocsf.md ---
    delegator: str | None = None
    delegatee: str | None = None
    type: str | None = None  # on_behalf_of, token_exchange, capability_grant, ...
    scope: list[str] = Field(default_factory=list)
    proof_type: str | None = None  # dpop, mtls_binding, signed_assertion
    ttl_seconds: int | None = None


class AgentProtocolID(IntEnum):
    """Agent-to-agent communication protocol (OCSF ``agent_message.protocol_id``).

    One generic discriminator across agentic protocols rather than a dedicated
    OCSF object per protocol — protocol-specific detail stays in the
    per-protocol OTel namespaces.
    """
    UNKNOWN = 0
    A2A = 1
    ACP = 2
    ANP = 3
    MCP = 4
    OTHER = 99


AGENT_PROTOCOL_LABELS: dict[int, str] = {
    0: "Unknown", 1: "A2A", 2: "ACP", 3: "ANP", 4: "MCP", 99: "Other",
}

_PROTOCOL_TO_ID: dict[str, int] = {
    "a2a": AgentProtocolID.A2A,
    "acp": AgentProtocolID.ACP,
    "anp": AgentProtocolID.ANP,
    "mcp": AgentProtocolID.MCP,
}


def normalize_agent_protocol_id(protocol: str | None) -> int:
    """Map a protocol string to an OCSF ``agent_message.protocol_id``."""
    if not protocol:
        return AgentProtocolID.UNKNOWN
    return int(_PROTOCOL_TO_ID.get(protocol.strip().lower(), AgentProtocolID.OTHER))


class OCSFAgentMessage(BaseModel):
    """OCSF ``agent_message`` object — one generic representation of an
    agent-to-agent communication across A2A / ACP / ANP / MCP.

    Proposed addition (see ocsf-mapping/ocsf-pr-draft.md). Carries the wire
    ``protocol_id`` discriminator plus the shared core (peer agents, unit of
    work + lifecycle status, transport, trust); protocol-specific extras live
    in ``metadata``.
    """
    protocol_id: int = AgentProtocolID.UNKNOWN
    protocol: str | None = None
    protocol_version: str | None = None
    direction: str | None = None  # request | response | stream | notification
    role: str | None = None       # client | server
    operation: str | None = None
    unit_uid: str | None = None
    unit_type: str | None = None  # task | run | message
    status: str | None = None     # canonical lifecycle status
    previous_status: str | None = None
    src_agent: "OCSFAIAgent | None" = None
    dst_agent: "OCSFAIAgent | None" = None
    delegation: "OCSFDelegation | None" = None
    parts_count: int | None = None
    part_types: list[str] = Field(default_factory=list)
    artifacts_count: int | None = None
    transport: str | None = None
    endpoint: str | None = None
    peer_endpoint: str | None = None
    trust_domain: str | None = None
    peer_trust_domain: str | None = None
    cross_domain: bool | None = None
    peer_did: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    duration_ms: float | None = None
    metadata: dict[str, Any] | None = None


class OCSFDelegationNode(BaseModel):
    """A single node in an OCSF ``delegation_lineage`` graph (OCSF issue #1640)."""
    uid: str
    parent_uid: str | None = None
    agent_uid: str | None = None
    depth: int | None = None


class OCSFDelegationLineage(BaseModel):
    """OCSF ``delegation_lineage`` — directed graph for ancestry queries."""
    nodes: list[OCSFDelegationNode] = Field(default_factory=list)


class ComplianceMetadata(BaseModel):
    """Compliance framework mappings."""
    nist_ai_rmf: dict[str, Any] | None = None
    mitre_atlas: dict[str, Any] | None = None
    iso_42001: dict[str, Any] | None = None
    eu_ai_act: dict[str, Any] | None = None
    soc2: dict[str, Any] | None = None
    gdpr: dict[str, Any] | None = None
    ccpa: dict[str, Any] | None = None
    csa_aicm: dict[str, Any] | None = None


# --- OCSF Base Event ---

class AIBaseEvent(BaseModel):
    """Base OCSF event for AITF AI events.

    Subclasses set ``category_uid`` and ``class_uid`` to the OCSF class they
    reuse (OCSF PR #1641 / issue #1640). AI-specific context is carried on the
    ``ai_operation`` profile (``ai_agent``, ``ai_model``, ``delegation``).
    """
    activity_id: int = OCSFActivity.OTHER
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int
    type_uid: int = 0
    time: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    severity_id: int = OCSFSeverity.INFORMATIONAL
    status_id: int = OCSFStatus.SUCCESS
    message: str = ""
    metadata: OCSFMetadata = Field(default_factory=OCSFMetadata)
    actor: OCSFActor | None = None
    device: OCSFDevice | None = None
    compliance: ComplianceMetadata | None = None
    observables: list[OCSFObservable] = Field(default_factory=list)
    enrichments: list[OCSFEnrichment] = Field(default_factory=list)

    # OCSF ``ai_operation`` profile (OCSF PR #1641) + delegation context
    # (OCSF issue #1640). Populated by the crosswalk so every AITF event can
    # be attributed to the AI agent and delegation that produced it.
    ai_agent: OCSFAIAgent | None = None
    ai_model: str | None = None
    delegation: OCSFDelegation | None = None
    delegation_lineage: OCSFDelegationLineage | None = None

    @model_validator(mode="after")
    def compute_type_uid(self) -> "AIBaseEvent":
        if self.type_uid == 0:
            self.type_uid = self.class_uid * 100 + self.activity_id
        return self

File: telemetry/aitf/sdk/typescript/src/ocsf/schema.ts
/**
 * AITF OCSF Base Schema.
 *
 * OCSF v1.9.0 base objects and AI-specific extension models.
 *
 * OCSF v1.9.0 (released 2026-08-03) landed the `ai_agent` and `delegation`
 * objects and the `ai_operation` profile, and attached that profile to the
 * `system`, `network`, `application` and `iam` base classes. AITF therefore
 * emits AI telemetry on released OCSF classes carrying `ai_operation`, and
 * no longer proposes a dedicated `ai` category — OCSF `categories.json`
 * stops at uid 8 (`unmanned_systems`).
 */

import { randomUUID } from "crypto";

// --- OCSF Enumerations ---

export enum OCSFSeverity {
  UNKNOWN = 0,
  INFORMATIONAL = 1,
  LOW = 2,
  MEDIUM = 3,
  HIGH = 4,
  CRITICAL = 5,
  FATAL = 6,
}

export enum OCSFStatus {
  UNKNOWN = 0,
  SUCCESS = 1,
  FAILURE = 2,
  OTHER = 99,
}

export enum OCSFActivity {
  UNKNOWN = 0,
  CREATE = 1,
  READ = 2,
  UPDATE = 3,
  DELETE = 4,
  OTHER = 99,
}

/**
 * OCSF category UIDs that AITF AI events map onto.
 *
 * Following OCSF's "reuse existing objects and profiles" approach — which
 * OCSF v1.9.0 ratified — AITF emits all AI telemetry under existing OCSF
 * categories enriched with the `ai_operation` profile.
 *
 * There is deliberately no `AI` member. OCSF `categories.json` stops at
 * uid 8 (`unmanned_systems`); a dedicated AI category was never ratified,
 * so AITF does not emit one. See `LEGACY_AI_CLASS_UIDS`.
 */
export enum OCSFCategoryUID {
  FINDINGS = 2,
  IAM = 3,
  DISCOVERY = 5,
  APPLICATION = 6,
}

/**
 * OCSF event class UIDs that AITF AI events map onto.
 *
 * Every UID below is a released OCSF class. Control-plane lifecycle
 * (agent, delegation, agent-to-agent communication) reuses released classes
 * carrying the `ai_operation` profile rather than bespoke 9xxx classes.
 */
export enum OCSFClassUID {
  // Reused existing OCSF classes (verified against OCSF v1.9.0).
  VULNERABILITY_FINDING = 2002,
  COMPLIANCE_FINDING = 2003,
  DETECTION_FINDING = 2004,
  ACCOUNT_CHANGE = 3001,
  AUTHENTICATION = 3002,
  AUTHORIZE_SESSION = 3003,
  ENTITY_MANAGEMENT = 3004,
  USER_ACCESS_MANAGEMENT = 3005,
  INVENTORY_INFO = 5001,
  WEB_RESOURCES_ACTIVITY = 6001,
  APPLICATION_LIFECYCLE = 6002,
  API_ACTIVITY = 6003,
  DATASTORE_ACTIVITY = 6005,
}

/**
 * Decode table for telemetry recorded by AITF <= 0.4, when AITF proposed a
 * dedicated `ai` category (uid 9) with classes 9001-9003. OCSF v1.9.0 shipped
 * the `ai_operation` profile on released base classes instead, so those UIDs
 * are retired. Retained so historical events remain decodable.
 */
export const LEGACY_AI_CLASS_UIDS: Record<number, OCSFClassUID> = {
  9001: OCSFClassUID.API_ACTIVITY, // agent_activity
  9002: OCSFClassUID.AUTHORIZE_SESSION, // delegation_activity
  9003: OCSFClassUID.API_ACTIVITY, // agent_communication
};

/**
 * Backward-compatible alias. AITF previously defined a bespoke Category 7 with
 * classes 7001-7010; events now reuse the OCSF classes above per OCSF's
 * object/profile-reuse model. Kept so existing imports keep working.
 */
export const AIClassUID = OCSFClassUID;

/**
 * OCSF `ai_agent.type_id` — normalized agent framework.
 *
 * Mirrors the enum introduced by OCSF PR #1641 (`objects/ai_agent.json`)
 * so AITF telemetry maps cleanly onto the upstream OCSF `ai_agent` object.
 */
export enum AgentTypeID {
  UNKNOWN = 0,
  NATIVE = 1,
  LANGCHAIN = 2,
  AUTOGEN = 3,
  CREWAI = 4,
  OTHER = 99,
}

/**
 * Caption labels for AgentTypeID.
 *
 * Matches the `ai_agent.type_id` enum as released in OCSF v1.9.0. Note the
 * `MCP` and `A2A` members discussed on OCSF PR #1641 did not ship; agentic
 * protocols are carried by `AgentProtocolID` below instead.
 */
export const AGENT_TYPE_LABELS: Record<number, string> = {
  0: "Unknown",
  1: "Native",
  2: "LangChain",
  3: "AutoGen",
  4: "CrewAI",
  99: "Other",
};

/**
 * AITF framework value -> OCSF ai_agent.type_id. Frameworks without a
 * dedicated OCSF enum member (langgraph, semantic_kernel, custom, ...)
 * normalize to OTHER (99), matching OCSF's open-enum guidance.
 */
const FRAMEWORK_TO_TYPE_ID: Record<string, AgentTypeID> = {
  native: AgentTypeID.NATIVE,
  langchain: AgentTypeID.LANGCHAIN,
  langgraph: AgentTypeID.LANGCHAIN,
  autogen: AgentTypeID.AUTOGEN,
  crewai: AgentTypeID.CREWAI,
};

/** Map an AITF framework string to an OCSF `ai_agent.type_id` value. */
export function normalizeAgentTypeId(framework?: string | null): number {
  if (!framework) {
    return AgentTypeID.UNKNOWN;
  }
  const key = framework.trim().toLowerCase();
  return key in FRAMEWORK_TO_TYPE_ID
    ? FRAMEWORK_TO_TYPE_ID[key]
    : AgentTypeID.OTHER;
}

/**
 * Retired. AITF once proposed a dedicated `ai` category (uid 9) under OCSF
 * issue #1640; OCSF v1.9.0 ratified the profile-on-existing-classes approach
 * instead and `categories.json` stops at uid 8. Exported as `null` rather
 * than deleted so any caller still reading it fails loudly instead of
 * silently emitting an unratified category.
 */
export const OCSF_AI_CATEGORY_UID: number | null = null;

// --- OCSF Base Object Interfaces ---

/** OCSF event metadata. */
export interface OCSFMetadata {
  version: string;
  product: {
    name: string;
    vendor_name: string;
    version: string;
  };
  uid: string;
  correlation_uid?: string;
  original_time?: string;
  logged_time: string;
}

/** OCSF actor information. */
export interface OCSFActor {
  user?: Record<string, unknown>;
  session?: Record<string, unknown>;
  app_name?: string;
}

/** OCSF device/host information. */
export interface OCSFDevice {
  hostname?: string;
  ip?: string;
  type?: string;
  os?: Record<string, string>;
  cloud?: Record<string, string>;
  container?: Record<string, string>;
}

/** OCSF enrichment data. */
export interface OCSFEnrichment {
  name: string;
  value: string;
  type?: string;
  provider?: string;
}

/** OCSF observable value. */
export interface OCSFObservable {
  name: string;
  type: string;
  value: string;
}

// --- AI-Specific Extension Interfaces ---

/** AI model information. */
export interface AIModelInfo {
  model_id: string;
  name?: string;
  version?: string;
  provider?: string;
  type?: string; // llm, embedding, image, audio, multimodal
  parameters?: Record<string, unknown>;
}

/** AI token usage statistics. */
export interface AITokenUsage {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  cached_tokens: number;
  reasoning_tokens: number;
  estimated_cost_usd?: number;
}

/** AI operation latency metrics. */
export interface AILatencyMetrics {
  total_ms: number;
  time_to_first_token_ms?: number;
  tokens_per_second?: number;
  queue_time_ms?: number;
  inference_time_ms?: number;
}

/** AI operation cost information. */
export interface AICostInfo {
  input_cost_usd: number;
  output_cost_usd: number;
  total_cost_usd: number;
  currency: string;
}

/** Multi-agent team information. */
export interface AITeamInfo {
  team_name: string;
  team_id?: string;
  topology?: string;
  members: string[];
  coordinator?: string;
}

/** Security finding details. */
export interface AISecurityFinding {
  finding_type: string;
  owasp_category?: string;
  risk_level: string;
  risk_score: number;
  confidence: number;
  detection_method: string;
  blocked: boolean;
  details?: string;
  pii_types: string[];
  matched_patterns: string[];
  remediation?: string;
}

/**
 * OCSF `ai_agent` object — **released in OCSF v1.9.0**.
 *
 * All eight fields below are upstream attributes; nothing here is an AITF
 * extension. An autonomous AI agent operating under delegated authority,
 * distinct from
 * the OCSF `agent` object (which models security sensors such as EDR/DLP)
 * and from human principals. Attached to events via the `ai_operation`
 * profile so any activity can be attributed to the agent that performed it.
 */
export interface OCSFAIAgent {
  uid: string; // required: stable logical identifier
  instance_uid?: string; // restart-sensitive running instance id
  name?: string;
  type?: string; // caption of type_id (Native, LangChain, ...)
  type_id: number;
  ai_model?: string; // model backing the agent at event time
  version?: string; // agent code/configuration revision
  charter?: string; // role / operating-boundary reference
}

/**
 * OCSF `delegation` object — **released in OCSF v1.9.0**.
 *
 * A durable authorization context that persists independently of any single
 * trace or session. The first four fields are the released upstream
 * attributes. The rest are an **AITF extension** and are the remaining OCSF
 * ask: upstream models *that* a delegation exists and who issued it, but not
 * what authority it conveys. See `upstream-pr-plan-ocsf.md`.
 */
export interface OCSFDelegation {
  // --- released OCSF v1.9.0 attributes ---
  uid: string; // required: stable delegation identifier
  created_time?: string; // when the delegation was minted
  parent_uid?: string; // parent delegation (lineage)
  issuer_uid?: string; // trusted issuer that minted the delegation
  // --- AITF extension: pending upstream ---
  delegator?: string;
  delegatee?: string;
  type?: string; // on_behalf_of, token_exchange, capability_grant, ...
  scope: string[];
  proof_type?: string; // dpop, mtls_binding, signed_assertion
  ttl_seconds?: number;
}

/** A single node in an OCSF `delegation_lineage` graph (OCSF issue #1640). */
export interface OCSFDelegationNode {
  uid: string;
  parent_uid?: string;
  agent_uid?: string;
  depth?: number;
}

/** OCSF `delegation_lineage` — directed graph for ancestry queries. */
export interface OCSFDelegationLineage {
  nodes: OCSFDelegationNode[];
}

/**
 * Agent-to-agent communication protocol (OCSF `agent_message.protocol_id`).
 *
 * One generic discriminator across agentic protocols rather than a dedicated
 * OCSF object per protocol — protocol-specific detail stays in the
 * per-protocol OTel namespaces.
 */
export enum AgentProtocolID {
  UNKNOWN = 0,
  A2A = 1,
  ACP = 2,
  ANP = 3,
  MCP = 4,
  OTHER = 99,
}

export const AGENT_PROTOCOL_LABELS: Record<number, string> = {
  0: "Unknown",
  1: "A2A",
  2: "ACP",
  3: "ANP",
  4: "MCP",
  99: "Other",
};

const PROTOCOL_TO_ID: Record<string, AgentProtocolID> = {
  a2a: AgentProtocolID.A2A,
  acp: AgentProtocolID.ACP,
  anp: AgentProtocolID.ANP,
  mcp: AgentProtocolID.MCP,
};

/** Map a protocol string to an OCSF `agent_message.protocol_id`. */
export function normalizeAgentProtocolId(protocol?: string | null): number {
  if (!protocol) {
    return AgentProtocolID.UNKNOWN;
  }
  const key = protocol.trim().toLowerCase();
  return key in PROTOCOL_TO_ID ? PROTOCOL_TO_ID[key] : AgentProtocolID.OTHER;
}

/**
 * OCSF `agent_message` object — one generic representation of an
 * agent-to-agent communication across A2A / ACP / ANP / MCP.
 *
 * Carries the wire `protocol_id` discriminator plus the shared core (peer
 * agents, unit of work + lifecycle status, transport, trust); protocol-specific
 * extras live in `metadata`.
 */
export interface OCSFAgentMessage {
  protocol_id: number;
  protocol?: string;
  protocol_version?: string;
  direction?: string; // request | response | stream | notification
  role?: string; // client | server
  operation?: string;
  unit_uid?: string;
  unit_type?: string; // task | run | message
  status?: string; // canonical lifecycle status
  previous_status?: string;
  src_agent?: OCSFAIAgent;
  dst_agent?: OCSFAIAgent;
  delegation?: OCSFDelegation;
  parts_count?: number;
  part_types: string[];
  artifacts_count?: number;
  transport?: string;
  endpoint?: string;
  peer_endpoint?: string;
  trust_domain?: string;
  peer_trust_domain?: string;
  cross_domain?: boolean;
  peer_did?: string;
  error_code?: string;
  error_message?: string;
  duration_ms?: number;
  metadata?: Record<string, unknown>;
}

/** Compliance framework mappings. */
export interface ComplianceMetadata {
  nist_ai_rmf?: Record<string, unknown>;
  mitre_atlas?: Record<string, unknown>;
  iso_42001?: Record<string, unknown>;
  eu_ai_act?: Record<string, unknown>;
  soc2?: Record<string, unknown>;
  gdpr?: Record<string, unknown>;
  ccpa?: Record<string, unknown>;
  csa_aicm?: Record<string, unknown>;
}

// --- OCSF Base Event ---

/**
 * Base OCSF event for AITF AI events.
 *
 * Subclasses/factories set `category_uid` and `class_uid` to the OCSF class
 * they reuse (OCSF PR #1641 / issue #1640). AI-specific context is carried on
 * the `ai_operation` profile (`ai_agent`, `ai_model`, `delegation`).
 */
export interface AIBaseEvent {
  activity_id: number;
  category_uid: number;
  class_uid: number;
  type_uid: number;
  time: string;
  severity_id: number;
  status_id: number;
  message: string;
  metadata: OCSFMetadata;
  actor?: OCSFActor;
  device?: OCSFDevice;
  compliance?: ComplianceMetadata;
  observables: OCSFObservable[];
  enrichments: OCSFEnrichment[];

  // OCSF `ai_operation` profile (OCSF PR #1641) + delegation context
  // (OCSF issue #1640). Populated by the crosswalk so every AITF event can
  // be attributed to the AI agent and delegation that produced it.
  ai_agent?: OCSFAIAgent;
  ai_model?: string;
  delegation?: OCSFDelegation;
  delegation_lineage?: OCSFDelegationLineage;
}

// --- Factory Functions ---

/** Create default OCSF metadata. */
export function createMetadata(
  correlationUid?: string
): OCSFMetadata {
  return {
    version: "1.9.0",
    product: {
      name: "AITF",
      vendor_name: "AITF",
      version: "0.4.0",
    },
    uid: randomUUID(),
    correlation_uid: correlationUid,
    logged_time: new Date().toISOString(),
  };
}

/** Create default token usage. */
export function createTokenUsage(
  options: Partial<AITokenUsage> = {}
): AITokenUsage {
  const usage: AITokenUsage = {
    input_tokens: options.input_tokens ?? 0,
    output_tokens: options.output_tokens ?? 0,
    total_tokens: options.total_tokens ?? 0,
    cached_tokens: options.cached_tokens ?? 0,
    reasoning_tokens: options.reasoning_tokens ?? 0,
    estimated_cost_usd: options.estimated_cost_usd,
  };
  if (usage.total_tokens === 0) {
    usage.total_tokens = usage.input_tokens + usage.output_tokens;
  }
  return usage;
}

/** Create a base OCSF AI event. */
export function createBaseEvent(
  classUid: number,
  options: Partial<AIBaseEvent> = {}
): AIBaseEvent {
  const activityId = options.activity_id ?? OCSFActivity.OTHER;
  return {
    activity_id: activityId,
    // Default to APPLICATION (6); factories override with the reused OCSF
    // category for the class they emit (OCSF PR #1641 / issue #1640).
    category_uid: options.category_uid ?? OCSFCategoryUID.APPLICATION,
    class_uid: classUid,
    type_uid: options.type_uid ?? classUid * 100 + activityId,
    time: options.time ?? new Date().toISOString(),
    severity_id: options.severity_id ?? OCSFSeverity.INFORMATIONAL,
    status_id: options.status_id ?? OCSFStatus.SUCCESS,
    message: options.message ?? "",
    metadata: options.metadata ?? createMetadata(),
    actor: options.actor,
    device: options.device,
    compliance: options.compliance,
    observables: options.observables ?? [],
    enrichments: options.enrichments ?? [],
  };
}

/**
 * Remove undefined/null properties from an object (for JSON export).
 */
export function stripNulls<T extends Record<string, unknown>>(
  obj: T
): Partial<T> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    if (value !== undefined && value !== null) {
      if (
        typeof value === "object" &&
        !Array.isArray(value) &&
        value !== null
      ) {
        const stripped = stripNulls(value as Record<string, unknown>);
        if (Object.keys(stripped).length > 0) {
          result[key] = stripped;
        }
      } else {
        result[key] = value;
      }
    }
  }
  return result as Partial<T>;
}
